### PR TITLE
Ensure Nested Loop Join output follows probe order

### PR DESCRIPTION
**Pull request by pedroerp**
Summary:
To follow expectations from engines such as Presto, changing NLJ
operator to emit output in the same order as probe inputs are read. This
required us to make changes to how the operator processes data internally
(check the documentation added to the header file).
.
The main difference (other than a large code refactor), is the now each probe
record needs to be processed entirely before we can move to the next; this
means we can emit probe mismatches right away in the right order.
.
As a downside, since we process probe records entirely, the output may contain
records from multiple build vectors, so we can't just wrap them into
dictionaries. We still produce dictionaries wrapped around probe columns
though.

Differential Revision: D60685613

### CODE REVIEW

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 1]
```diff
@@ -162,136 +242,246 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
...
+    const RowVectorPtr& buildVector,
+    vector_size_t buildRow) {
+  // Probe side is always a dictionary; just populate the indice.
+  rawProbeIndices_[numOutputRows_] = probeRow_;
+
+  // Copy the build projections.
+  for (const auto& projection : buildProjections_) {
+    const auto& buildChild = buildVector->childAt(projection.inputChannel);
+    const auto& outputChild = output_->childAt(projection.outputChannel);
+    outputChild->copy(buildChild.get(), numOutputRows_, buildRow, 1);
```
Comment (Yuhta) [inline velox/exec/NestedLoopJoinProbe.cpp:446]: This will kill the performance.  Can we accumulate `buildRow`s as `CopyRange`s or `toSourceRow` and only call `copy` once per output batch?

**Comment by Yuhta** [comment]
Also maybe we should consider merging build side vectors into one large vector, this way we would be able to wrap them in dictionary.  Is there any concern about this?

**Comment by pedroerp** [comment]
> Also maybe we should consider merging build side vectors into one large vector, this way we would be able to wrap them in dictionary. Is there any concern about this?

We discussed this offline. Depending on the size of the build it may be tricky to allocate a large enough contiguous memory location. We decided to move on with this approach for now. I'll take a look at batching the copies like you suggested.

**Comment by pedroerp** [comment]
Just added the optimization suggested by @Yuhta . The code now keeps track of the ranges of rows from the build side to be copied to the output (merging them when possible), then copying them to the output using `BaseVector::copyRanges()` column-by-column. 

Another potential optimization would be to figure out if only copies from the same build buffer are need for an output buffer, we could create a dictionary instead. But it's not clear how frequently this would happen. Could be done as a follow up.

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 2]
```diff
@@ -77,6 +78,50 @@ void NestedLoopJoinProbe::initialize() {
   joinNode_.reset();
 }
 
+void NestedLoopJoinProbe::initializeFilter(
```
Comment (pedroerp) [inline velox/exec/NestedLoopJoinProbe.cpp:81]: just moved this function up to make the code order more readable; no changes here.

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 3]
```diff
@@ -132,9 +180,37 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
     probeSideEmpty_ = false;
   }
   VELOX_CHECK_EQ(buildIndex_, 0);
-  if (needsProbeMismatch(joinType_)) {
-    probeMatched_.resizeFill(input_->size(), false);
+}
+
+void NestedLoopJoinProbe::noMoreInput() {
```
Comment (pedroerp) [inline velox/exec/NestedLoopJoinProbe.cpp:185]: same for these two functions.

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 4]
```diff
@@ -162,136 +242,260 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
...
-                                                   input_,
-                                                   probeMatched_,
-                                                   probeOutMapping_,
-                                                   identityProjections_,
-                                                   buildProjections_)
-                                             : nullptr;
-      finishProbeInput();
-      break;
-    }
+    // Generate actual join input by processing probe and build matches, and
```
Comment (pedroerp) [inline velox/exec/NestedLoopJoinProbe.cpp:250]: from now on the changes start

**Comment by amitkdutta** [comment]
@pedroerp Thanks a lot Pedro for priortizing and fixing this long lasting issue. Recently @kgpai made a PR to disbale NLJ in Prestissimo worker (https://github.com/prestodb/presto/pull/23341), it will be great to run the e2e tests with Prestissimo to see if queries are working expected. This will test this PR well and advancing velox in PrestoDb won't create any unnecessary issues. CC: @aditi-pandit @majetideepak

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 5]
```diff
@@ -77,6 +78,50 @@ void NestedLoopJoinProbe::initialize() {
...
+  VELOX_CHECK_NULL(joinCondition_);
+
+  std::vector<core::TypedExprPtr> filters = {filter};
+  joinCondition_ =
+      std::make_unique<ExprSet>(std::move(filters), operatorCtx_->execCtx());
+
+  column_index_t filterChannel = 0;
+  std::vector<std::string> names;
+  std::vector<TypePtr> types;
+  auto numFields = joinCondition_->expr(0)->distinctFields().size();
```
Comment (aditi-pandit) [inline velox/exec/NestedLoopJoinProbe.cpp:94]: Nit : const

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 6]
```diff
@@ -77,6 +78,50 @@ void NestedLoopJoinProbe::initialize() {
...
+  names.reserve(numFields);
+  types.reserve(numFields);
+
+  for (auto& field : joinCondition_->expr(0)->distinctFields()) {
```
Comment (aditi-pandit) [inline velox/exec/NestedLoopJoinProbe.cpp:98]: Nit : const

**Review by aditi-pandit — COMMENTED** [review]
Thanks @pedroerp for the code and detailed comments.

Had some very minor comments.

**Comment by kgpai** [comment]
@aditi-pandit The pr for disabling NLJ has not been merged, and if this is ready its not required.

> https://github.com/prestodb/presto/pull/23315/files#diff-a07cf9ace74d4a8c17fce624676de4770bbcf47547e69d35c70d12777d57d9a1

Would be good to add this in a subsequent Prestissimo PR upgrading Velox.

File: velox/exec/NestedLoopJoinProbe.h [hunk 7]
```diff
@@ -20,6 +20,39 @@
...
+
+/// This class implements a Nested Loop Join (NLJ) between records from the
+/// probe (input_) and build (NestedLoopJoinBridge) sides. It supports inner,
+/// left, right and full outer joins.
+///
+/// This class is generally useful to evaluate non-equi-joins (e.g. "k1 >= k2"),
+/// when join conditions may need to be evaluated against a full cross product
+/// of the input.
+///
+/// The output follows the order of the build side rows (for inner and left
```
[the remark this answers is not in the document]
Comment (pedroerp) [reply velox/exec/NestedLoopJoinProbe.h:32]: Good catch, I indeed meant "probe" here. :)

File: velox/exec/NestedLoopJoinProbe.cpp (hunk 4 above)
Comment (mbasmanova) [inline velox/exec/NestedLoopJoinProbe.cpp:250]: typo: input -> output ?

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 8]
```diff
@@ -162,136 +242,260 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
...
+        numOutputRows, probeRow_, input_->childAt(inputChannel));
   }
+
   return std::make_shared<RowVector>(
-      pool(), outputType_, nullptr, numUnmatched, std::move(projectedChildren));
+      pool(), outputType, nullptr, numOutputRows, std::move(projectedChildren));
+}
+
+void NestedLoopJoinProbe::addOutputRow(vector_size_t buildRow) {
+  // Probe side is always a dictionary; just populate the indice.
```
Comment (mbasmanova) [inline velox/exec/NestedLoopJoinProbe.cpp:439]: typo: indice -> index ?

File: velox/exec/NestedLoopJoinProbe.cpp [hunk 9]
```diff
@@ -162,136 +242,260 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
...
+      const auto& buildChild = buildVector->childAt(projection.inputChannel);
+      const auto& outputChild = output_->childAt(projection.outputChannel);
+      outputChild->copyRanges(buildChild.get(), buildCopyRanges_);
+    }
+    buildCopyRanges_.clear();
+  }
+}
+
+void NestedLoopJoinProbe::addProbeMismatchRow() {
+  // Probe side is always a dictionary; just populate the indice.
```
Comment (mbasmanova) [inline velox/exec/NestedLoopJoinProbe.cpp:465]: typo: indice -> index ?

File: velox/exec/NestedLoopJoinProbe.h [hunk 10]
```diff
@@ -20,6 +20,39 @@
 #include "velox/exec/ProbeOperatorState.h"
 
 namespace facebook::velox::exec {
+
+/// This class implements a Nested Loop Join (NLJ) between records from the
```
Comment (mbasmanova) [inline velox/exec/NestedLoopJoinProbe.h:24]: nit: drop "This class" and start with a verb "Implements ..."

File: velox/exec/NestedLoopJoinProbe.h (hunk 7 above)
Comment (mbasmanova) [inline velox/exec/NestedLoopJoinProbe.h]: build -> probe

File: velox/exec/tests/utils/QueryAssertions.cpp [hunk 11]
```diff
@@ -886,7 +886,11 @@ std::vector<MaterializedRow> materialize(const RowVectorPtr& vector) {
     MaterializedRow row;
     row.reserve(numColumns);
     for (size_t j = 0; j < numColumns; ++j) {
-      row.push_back(variantAt(vector->childAt(j), i));
+      // variantAt() assumes you can upcast to SimpleVector, so we need to take
+      // the inner vector out of lazies first.
+      const auto& child = vector->childAt(j);
+      row.push_back(variantAt(
+          child->isLazy() ? BaseVector::loadedVectorShared(child) : child, i));
```
Comment (mbasmanova) [inline velox/exec/tests/utils/QueryAssertions.cpp:893]: Can this change be extracted into a separate PR? 

Can we call loadedVectorShared once per column vs. once per column per row?

**Review by mbasmanova — COMMENTED** [review]
It would be nice to update documentation in https://facebookincubator.github.io/velox/develop/operators.html#nestedloopjoinnode and in PlanNode.h

**Comment by amitkdutta** [comment]
> > @pedroerp Thanks a lot Pedro for priortizing and fixing this long lasting issue. Recently @kgpai made a PR to disbale NLJ in Prestissimo worker ([prestodb/presto#23341](https://github.com/prestodb/presto/pull/23341)), it will be great to run the e2e tests with Prestissimo to see if queries are working expected. This will test this PR well and advancing velox in PrestoDb won't create any unnecessary issues. CC: @aditi-pandit @majetideepak
> 
> @pedroerp : Agree. Would be great to re-enable NLJ and test all the queries in presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeGeneralQueries.java in https://github.com/prestodb/presto/pull/23315/files#diff-a07cf9ace74d4a8c17fce624676de4770bbcf47547e69d35c70d12777d57d9a1

Confirming that we run the e2e tests in meta internal tests and they are all good in this PR.

File: velox/core/PlanNode.h [hunk 12]
```diff
@@ -1669,7 +1669,8 @@ class MergeJoinNode : public AbstractJoinNode {
 ///
 /// Nested loop join supports both equal and non-equal joins. Expressions
 /// specified in joinCondition are evaluated on every combination of left/right
-/// tuple, to emit result.
+/// tuple, to emit result. Results are emitted following the same input order of
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1672]: It would be nice to also update documentation in https://facebookincubator.github.io/velox/develop/operators.html#nestedloopjoinnode

Should we clarify that the order is preserved only within a single thread of execution?

**Review by mbasmanova — APPROVED** [review]

**Comment by pedroerp** [comment]
Thank you @mbasmanova @Yuhta and @aditi-pandit for the quick code review. All comments are addressed. Will merge it later today unless I hear any last objections :)

**Review by aditi-pandit — APPROVED** [review]
Thanks @pedroerp.

I don't think there is much use of NLJ in TPC-DS queries, but we will do a run and monitor any perf degradations.

https://github.com/facebookincubator/velox/discussions/10343 describes the equivalent in Presto. We could consider implementing something like that in the future.

**Comment by amitkdutta** [comment]
> @aditi-pandit The pr for disabling NLJ has not been merged, and if this is ready its not required.
> 
> > https://github.com/prestodb/presto/pull/23315/files#diff-a07cf9ace74d4a8c17fce624676de4770bbcf47547e69d35c70d12777d57d9a1
> 
> Would be good to add this in a subsequent Prestissimo PR upgrading Velox.

@kgpai @pedroerp 
- Updated velox and merged in https://github.com/prestodb/presto/pull/23395. No errors in existing tests
- Copied tests from @aditi-pandit's https://github.com/prestodb/presto/pull/23315 which extensively tests this path in https://github.com/prestodb/presto/pull/23396

**Comment by pedroerp** [comment]
Thank you @amitkdutta !
